### PR TITLE
Add support to restrict @Core.AcceptableMediaTypes

### PR DESCRIPTION
**Pull request by samyuktaprabhu**
# Add Support to Restrict @Core.AcceptableMediaTypes

## ✨ New Features

Introduced support for restricting allowed MIME types for attachments using the `@Core.AcceptableMediaTypes` annotation. This feature enables validation of file types during upload at `HandlerOrder.BEFORE`, ensuring only specified media types are accepted before processing begins.

### Core Implementation

* **`CreateAttachmentsHandler.java`**: Added new `@Before` handler method `processBeforeForMetadata` that validates acceptable media types early in the request lifecycle. Updated constructor to accept `CdsRuntime` parameter for metadata validation support.

* **`AttachmentValidationHelper.java`** (new): Comprehensive helper class for media type validation with support for wildcard patterns (`image/*`) and default fallback behavior (`*/*`). Validates file types against configured acceptable media types.

* **`MediaTypeResolver.java`** (new): Extracts acceptable media types from CDS entity annotations, handling both direct media entities and composed attachments using the `AssociationCascader`.

* **`MediaTypeService.java`** (new): Provides MIME type resolution from file extensions using Java's `URLConnection.getFileNameMap()` and validates against acceptable media types with support for exact matches and wildcard patterns.

* **`AttachmentDataExtractor.java`** (new): Extracts and validates file names from attachment data using `CdsDataProcessor`, ensuring files have proper extensions before MIME type validation.

* **`Registration.java`**: Updated handler registration to pass `CdsRuntime` instance to `CreateAttachmentsHandler`.

* **`ApplicationHandlerHelper.java`**: Added `deepSearchForAttachments()` method functionality (refactored into `AssociationCascader`).

* **`AssociationCascader.java`**: Enhanced with `hasAttachmentPath()` and `findMediaEntityNames()` methods to recursively check for attachment compositions in entity hierarchies and collect media entity names.

### Documentation

* **`README.md`**: Added comprehensive documentation section "Restrict allowed MIME types" with examples of:
  - Exact matches (`image/jpeg`, `image/png`)
  - Wildcard patterns (`image/*`)
  - Default behavior (`*/*`)
  - Usage scenarios for single and multiple media types

### Test Suite

* **`CreateAttachmentsHandlerTest.java`**: Added tests for new metadata validation handler with proper annotation verification

* **`AttachmentValidationHelperTest.java`** (new): Comprehensive test suite with 16+ test cases covering MIME type detection, validation, wildcards, error scenarios, and edge cases including hidden files

* **`MediaTypeResolverTest.java`** (new): Tests for media type extraction from entity annotations

* **`MediaTypeServiceTest.java`** (new): Tests for MIME type resolution and validation logic including edge cases like hidden files (`.gitignore`)

* **`MediaValidatedAttachmentsDraftTest.java`** (new): Integration tests for draft service media validation with parameterized tests for various file types

* **`MediaValidatedAttachmentsNonDraftTest.java`** (new): Integration tests for non-draft service media validation, including deep create scenarios and multiple attachment validation

* **`AssociationCascaderTest.java`**: Added tests for new `hasAttachmentPath()` and `findMediaEntityNames()` methods

* **Size validation tests**: Updated to include `fileName` assignments for proper validation

### Integration Test Configuration

* **`data-model.cds`**: Added `mediaValidatedAttachments` and `mimeValidatedAttachments` compositions to Roots entity

* **`test-service.cds`**: Configured annotations to accept only JPEG and PNG images for `mediaValidatedAttachments`, and PDFs for `mimeValidatedAttachments`

* **`RootEntityBuilder.java`**: Enhanced builder to initialize `sizeLimitedAttachments` list in constructor for proper entity setup

### Sample Application

* **`samples/bookshop/srv/attachments.cds`**: 
  - Added `mediaValidatedAttachments` composition to Books entity
  - Configured to accept only JPEG and PNG images
  - Added UI annotations to hide the field from Browse Books App

- [ ] 🔄 Regenerate and Update Summary

---
📬 [Subscribe to the Hyperspace PR Bot DL](https://url.sap/451kgs) to get the latest announcements and pilot features!

<details>
<summary>PR Bot Information</summary>

**Version:** `1.17.80` | 📖 [Documentation](https://url.sap/dy9ocn) | 🚨 [Create Incident](https://url.sap/budnv9) | 💬 [Feedback](https://url.sap/my4dn3)

- LLM: `anthropic--claude-4.5-sonnet`
- Summary Prompt: [Default Prompt](https://github.tools.sap/intelligent-insights/i2-pull-request/blob/main/src/services/llm/prompts/summary_instructions_prompt.md)
- Output Template: [Default Template](https://github.tools.sap/intelligent-insights/i2-pull-request/blob/main/src/services/llm/prompts/summary_default_output_template.md)
- Event Trigger: `pull_request.edited`
- Correlation ID: `27d010e0-1609-11f1-9c18-5f2e249083ea`
</details>

### CODE REVIEW

File: cds-feature-attachments/src/test/java/com/sap/cds/feature/attachments/handler/applicationservice/helper/AttachmentValidationHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    assertEquals(ErrorStatuses.UNSUPPORTED_MEDIA_TYPE, ex.getErrorStatus());
+  }
+
+  @Test
+  void shouldReturnFalseWhenMimeTypeIsNull() {
+    boolean result = AttachmentValidationHelper.checkMimeTypeMatch(List.of("image/png"), null);
+
+    assertFalse(result);
+  }
+
```
Comment (Schmarvinius) [inline cds-feature-attachments/src/test/java/com/sap/cds/feature/attachments/handler/applicationservice/helper/AttachmentValidationHelperTest.java]: Maybe also add a test for wildcard subtype rejection like image/* should reject application/pdf

File: integration-tests/srv/src/test/java/com/sap/cds/feature/attachments/integrationtests/draftservice/MediaValidatedAttachmentsDraftTest.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                {
+                  "fileName": "%s"
+                }
+                """
+            .formatted(fileName);
+
+    requestHelper.executePostWithMatcher(
+        buildDraftAttachmentCreationUrl(rootId), metadata, status().is(expectedStatus));
+  }
+
```
Comment (Schmarvinius) [inline integration-tests/srv/src/test/java/com/sap/cds/feature/attachments/integrationtests/draftservice/MediaValidatedAttachmentsDraftTest.java]: Also add tests for false inputs like an empty string. You have them for non draft but not for draft.

File: integration-tests/srv/src/test/java/com/sap/cds/feature/attachments/integrationtests/nondraftservice/MediaValidatedAttachmentsNonDraftTest.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  void shouldValidateMediaTypes(String fileName, String mediaType, int expectedStatus)
+      throws Exception {
+    String rootId = createRootAndReturnId();
+    String attachmentMetadata = createAttachmentMetadata(fileName);
+
+    requestHelper.executePostWithMatcher(
+        createUrl(rootId, MEDIA_VALIDATED_ATTACHMENTS),
+        attachmentMetadata,
+        status().is(expectedStatus));
+  }
```
[the remark this answers is not in the document]
Comment (Schmarvinius) [reply integration-tests/srv/src/test/java/com/sap/cds/feature/attachments/integrationtests/nondraftservice/MediaValidatedAttachmentsNonDraftTest.java]: Again one integration test for dot files

File: integration-tests/srv/src/test/java/com/sap/cds/feature/attachments/integrationtests/draftservice/MediaValidatedAttachmentsDraftTest.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  // Helper methods
+  private String createDraftRootAndReturnId() throws Exception {
+    CdsData response =
+        requestHelper.executePostWithODataResponseAndAssertStatusCreated(BASE_ROOT_URL, "{}");
+
+    DraftRoots draftRoot = Struct.access(response).as(DraftRoots.class);
+
+    requestHelper.executePatchWithODataResponseAndAssertStatusOk(
+        getRootUrl(draftRoot.getId(), false),
+        """
```
Comment (Schmarvinius) [inline integration-tests/srv/src/test/java/com/sap/cds/feature/attachments/integrationtests/draftservice/MediaValidatedAttachmentsDraftTest.java]: Maybe use JSONObject and oonvert to string (cleaner)

File: cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/configuration/Registration.java [hunk 5]
```diff
@@ -89,7 +89,8 @@ public void eventHandlers(CdsRuntimeConfigurer configurer) {
           OutboxService.PERSISTENT_UNORDERED_NAME);
     }
 
-    // build malware scanner client, could be null if no service binding is available
+    // build malware scanner client, could be null if no service binding is
+    // available
```
Comment (Schmarvinius) [inline cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/configuration/Registration.java]:
```suggestion
    // build malware scanner client, could be null if no service binding is available
```

File: cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/configuration/Registration.java [hunk 6]
```diff
@@ -118,12 +119,14 @@ public void eventHandlers(CdsRuntimeConfigurer configurer) {
...
-    // register event handlers for application service, if at least one application service is
+    // register event handlers for application service, if at least one application
+    // service is
     // available
     boolean hasApplicationServices =
         serviceCatalog.getServices(ApplicationService.class).findFirst().isPresent();
     if (hasApplicationServices) {
-      configurer.eventHandler(new CreateAttachmentsHandler(eventFactory, storage, defaultMaxSize));
+      configurer.eventHandler(
+          new CreateAttachmentsHandler(eventFactory, storage, defaultMaxSize, runtime));
```
Comment (Schmarvinius) [inline cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/configuration/Registration.java]:
```suggestion
      configurer.eventHandler(new CreateAttachmentsHandler(eventFactory, storage, defaultMaxSize, runtime));
```

File: cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/configuration/Registration.java [hunk 7]
```diff
@@ -138,7 +141,8 @@ public void eventHandlers(CdsRuntimeConfigurer configurer) {
           "No application service is available. Application service event handlers will not be registered.");
     }
 
-    // register event handlers on draft service, if at least one draft service is available
+    // register event handlers on draft service, if at least one draft service is
+    // available
```
Comment (Schmarvinius) [inline cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/configuration/Registration.java]:
```suggestion
    // register event handlers on draft service, if at least one draft service is available
```

File: cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/handler/applicationservice/helper/validation/AttachmentValidationHelper.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    CdsModel cdsModel = cdsRuntime.getCdsModel();
+    Optional<CdsEntity> optionalServiceEntity = cdsModel.findEntity(entity.getQualifiedName());
+
+    if (optionalServiceEntity.isEmpty()) {
+      return;
+    }
+
+    CdsEntity serviceEntity = optionalServiceEntity.get();
+    boolean hasNoAttachmentCompositions =
+        !ApplicationHandlerHelper.deepSearchForAttachments(serviceEntity);
```
Comment (Schmarvinius) [inline cds-feature-attachments/src/main/java/com/sap/cds/feature/attachments/handler/applicationservice/helper/validation/AttachmentValidationHelper.java]: Instead use AssociationCascader and make use of existing code (deepSearchForAttachments was bad code from me back then and needs to be removed). This would also simplify the getAcceptableMediaTypesFromEntity function since you don't need to manually search for the paths because you already have the NodeTree.
